### ISSUE TITLE
When RAFT enabled, removenode operation failed while network partly blocked 

### ISSUE THREAD

**Issue by temichus** [scylladb/scylladb#12892]
## Issue description

- [ ] This issue is a regression.
- [x] It is unknown if this issue is a regression.

Dtest test case test_removenode_while_gossip_partly_blocked failed is case when RAFT enabled (-consistent-cluster-management flag). 
looks like  remove_node operation is stuck:
node 2 log:
```
INFO  2023-02-15 17:59:22,895 [shard 0] api - remove_node: host_id=8e99f750-a383-456a-9af1-fd5dc86cc63d ignore_nodes={}
INFO  2023-02-15 17:59:22,895 [shard 0] raft_group0 - Performing a group 0 read barrier...
ERROR 2023-02-15 18:14:31,761 [shard 0] rpc - client 127.0.42.2:7000: client connection dropped: recv: Connection timed out
INFO  2023-02-15 18:41:21,910 [shard 0] compaction_manager - Asked to stop
```
Dtest log:
```
17:59:22,263 209354  ccm                            DEBUG    cluster.py          :711  | test_removenode_while_gossip_partly_blocked: node2: nodetool cmd=removenode 8e99f750-a383-456a-9af1-fd5dc86cc63d wait=True timeout=None
18:41:21,890 209354  errors                         ERROR    conftest.py         :203  | test_removenode_while_gossip_partly_blocked: test failed: .....
......
 Failed: Timeout >2700.0s
```
at the same time **test passed when gossip enabled** 

## Impact

 remove_node operation is stuck:

## How frequently does it reproduce?
run command:
```
pytest --scylla-version='unstable/master:latest'  --consistent-cluster-management update_cluster_layout_tests.py::TestUpdateClusterLayout::test_removenode_while_gossip_partly_blocked
```

PR https://github.com/scylladb/scylla-dtest/pull/3041

*Installation details*
Scylla version (or git commit hash): 5.3.0~dev-0.20230214.2653865b34d8

url=http://s3.amazonaws.com/downloads.scylladb.com/unstable/scylla/master/relocatable/2023-02-14T08:49:47Z/scylla-unified-5.3.0~dev-0.20230214.2653865b34d8.x86_64.tar.gz

Cluster size: 3 ccm nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 20.04.5 LTS

Platform (physical/VM/cloud instance type/docker): Scylla CCM

logs
failed raft - https://drive.google.com/file/d/1UzzGOeVMfEDCKmOumTgISmyyNBiFH0Jw/view?usp=sharing
passed gossip - https://drive.google.com/file/d/1QFtM9BuQwSxcz_QtFA7eEgdhhIAlw5m1/view?usp=sharing

**Comment by mykaul** [comment]
@temichus - It would be good to indicate exact version used.

**Comment by kostja** [comment]
I don't see this test in dtest master. Could you please share here a link to the source file with the test?

**Comment by kostja** [comment]
@temichus

**Comment by mykaul** [comment]
> I figured it's a PR which is not merged yet. Please do not file a PR failure as a bug against Scylla master. The bug is more likely in the test.

They could share the branch, even before it's merged. What if indeed there IS a bug? It's unlikely they'll merge a test that fails (unless it's XFAIL) .

**Comment by kostja** [comment]
Why do you think it's a bug?

**Comment by kostja** [comment]
What is happening is that you try to commit a change to group0. This operation requires a majority, and under the hood in order to read group0 state the node performs a read barrier. This read barrier never finishes. removenode operation currently doesn't supply a timeout.I understand you appreciate the error message, but generally having an error message in raft based operation is an enhancement, and I think not a worthy one (better to add timeout support to removenode) not a bug.

**Comment by kostja** [comment]
> > I figured it's a PR which is not merged yet. Please do not file a PR failure as a bug against Scylla master. The bug is more likely in the test.
> 
> They could share the branch, even before it's merged. What if indeed there IS a bug? It's unlikely they'll merge a test that fails (unless it's XFAIL) .

I just got confused reading it's a dtest issue, so automatically assuming it's a regression of some sort.

**Comment by mykaul** [comment]
> What is happening is that you try to commit a change to group0. This operation requires a majority, and under the hood in order to read group0 state the node performs a read barrier. This read barrier never finishes. removenode operation currently doesn't supply a timeout.I understand you appreciate the error message, but generally having an error message in raft based operation is an enhancement, and I think not a worthy one (better to add timeout support to removenode) not a bug.

The overall user experience is inferior to what the user had before Raft. Somewhere along the stack (which may not be in Raft), I expect a timeout for the operation and an error returned. Extra points if it can be a clear error as we've had in the past, prior to Raft.

**Comment by kostja** [comment]
Just imagine you didn't have a toy installation with empty nodes but a real cluster and the network partitioning would happen during streaming. You would get exact same behaviour as you got with Raft - the process would get stuck. It's a courtesy of the current gossip implementation that it looks if a node is up or down and reports an error but it is just one check along the long sequence of operations a removenode performs.

And you can't expect it to timeout either, since it can do a lot of work over slow network, so there is no reasonable timeout.

**Comment by kostja** [comment]
I honestly don't know the UX which would be useful in a general case. The error you received in the test may be just undesirable in the circumstances of a disaster recovery during a storm as it seems nice in an empty cluster with an artificial network parittioning.

**Comment by kostja** [comment]
One way out could be to allow removenode to supply a user defined timeout. Then a DBA depending on the size of the cluster and expectations could set a high or low timeout. We'd be happy to support it in Raft, but I'm afraid a lot of streaming code will need to change to support it as well - it's a big enhancement @asias

**Comment by kostja** [comment]
Why is the test scenario isolating two nodes out of three then? It sould be enough to isolate the node that is being decommissioned.

**Comment by fruch** [comment]
> Why is the test scenario isolating two nodes out of three then? It should be enough to isolate the node that is being decommissioned. @temichus 

@kostja is a split case, 2 out of 3 and 1 out of 3, is the same.

The issue this test is reproducing cases we run into under k8s that nodes are restated and communication were blocked only in one direction, which in turn made a node to be decommissioned without the other nodes getting that via gossip

The fix for it was also touching code path touching remove node action, and we wanted to check the effect on it too

**Comment by fruch** [comment]
> @temichus - It would be good to indicate exact version used.

It's hidden in the url mentioned that in the description

**Comment by fruch** [comment]
> 
> I just got confused reading it's a dtest issue, so automatically assuming it's a regression of some sort.

So it's agreed that this should be reopened?

**Comment by kostja** [comment]
> @kostja is a split case, 2 out of 3 and 1 out of 3, is the same.

It's not the same if you run with raft.

**Comment by kostja** [comment]
> > I just got confused reading it's a dtest issue, so automatically assuming it's a regression of some sort.
> 
> So it's agreed that this should be reopened?

I think you should reopen the original bug that fixes the original test case you're adding a test for. It should be working with Raft. 

The sequence of events "report an issue", "provide a partial fix in gossip only", "write a test that happens to showcase the fix is working in one scenario, but doesn't test more realistic scenarios", "open a bug against raft mode" is not going to have the issue fixed.

Please note that I wrote above that there is no reliable way to provide the expected behavior. Please address that remark. How you expect this to be fixed? I would not want a fix that masks the real problem and just passes the test case provided.

**Comment by kostja** [comment]
> > What is happening is that you try to commit a change to group0. This operation requires a majority, and under the hood in order to read group0 state the node performs a read barrier. This read barrier never finishes. removenode operation currently doesn't supply a timeout.I understand you appreciate the error message, but generally having an error message in raft based operation is an enhancement, and I think not a worthy one (better to add timeout support to removenode) not a bug.
> 
> The overall user experience is inferior to what the user had before Raft. Somewhere along the stack (which may not be in Raft), I expect a timeout for the operation and an error returned. Extra points if it can be a clear error as we've had in the past, prior to Raft.

@mykaul I 

> > What is happening is that you try to commit a change to group0. This operation requires a majority, and under the hood in order to read group0 state the node performs a read barrier. This read barrier never finishes. removenode operation currently doesn't supply a timeout.I understand you appreciate the error message, but generally having an error message in raft based operation is an enhancement, and I think not a worthy one (better to add timeout support to removenode) not a bug.
> 
> The overall user experience is inferior to what the user had before Raft. Somewhere along the stack (which may not be in Raft), I expect a timeout for the operation and an error returned. Extra points if it can be a clear error as we've had in the past, prior to Raft.

I'd agree with this remark if it was a regression, but this functionality was added as part of RBNO. 

RBNO, apart from switching to repair-based streaming, adds multiple safety fixes, like this one,  but they are best effort. You can consult with the code in service/storage_service.cc, the error message shown in the "expected" out is a result of a single-place check at start of RBNO code path. If the network partition happens to happen a moment later, you're going to get the same behaviour as with Raft.

Honestly, I could just accept this as a bug bug and copy paste the check to the raft branch of execution. But this would not address the underlying cause.

**Comment by roydahan** [comment]
@kostja regardless of to reopen https://github.com/scylladb/scylladb/issues/11302, we will continue this discussion on an opened issue.

**Comment by kostja** [comment]
> > > > I just got confused reading it's a dtest issue, so automatically assuming it's a regression of some sort.
> > > 
> > > 
> > > So it's agreed that this should be reopened?
> > 
> > 
> > I think you should reopen the original bug that fixes the original test case you're adding a test for. It should be working with Raft.
> > The sequence of events "report an issue", "provide a partial fix in gossip only", "write a test that happens to showcase the fix is working in one scenario, but doesn't test more realistic scenarios", "open a bug against raft mode" is not going to have the issue fixed.
> > Please note that I wrote above that there is no reliable way to provide the expected behavior. Please address that remark. How you expect this to be fixed? I would not want a fix that masks the real problem and just passes the test case provided.
> 
> I don't understand, this is a real case happened for us under k8s, and you want me to define how it should be addressed in scylla ?

A bug is a program behavior contradicting the documentation or user expectations. The program did not behave like that before RBNO, RBNO set the expectations here. They look intuitive and nice, but they are not realistic. Given that this change has never been the purpose of RBNO, it is a by no means nice, but separate imrpovement, do you think it is correct to refer to the absence of this improvement on the raft branch of the code which forked off the master pretty much in parallel to rbno as a bug? This feature is simply missing in Raft mode.

> If you argue we need more nodes in this case in Raft that's one thing, but plainly saying this is not a valid issue cause you don't think the fix done in gossip is good enough...

I argue it's not valid because you're referring to the fix done in gossip as reference behavior. I am explaining that this reference behavior is not realistic in all scenarios, which a more extensive test for this problem would notice.

**Comment by fruch** [comment]
> user expectations

expectation as a user is that something that is working, to keep working.
also as a user I don't expect any operation I do to hang forever with no indication why.

Again, you might be correct about that the fix that was done isn't good enough, and think it's bigger work then a "fix", and it's a feature, title it as you wish. also no one titled this a P1

**Comment by kostja** [comment]
> > user expectations
> 
> expectation as a user is that something that is working, to keep working. 

I wrote above why I don't think it's "working": 1) it was not "working" before RBNO 2) it is not "working" unless your spaces are empty and you are kind enough to wait for gossip to detect the network failure. 

This window of "working" is so narrow, that you had to carefully devise a test to detect it. 

> also as a user I don't expect any operation I do to hang forever with no indication why.

Nice expectation, this is why there is a feature request that nodetool operations should report progress.
> 
> Again, you might be correct about that the fix that was done isn't good enough, and think it's bigger work then a "fix", and it's a feature, title it as you wish. also no one titled this a P1

If that is the case then the issues database is just a wrong venue for it. As a godzillion of comments on this issue already indicate. We need a PRD where we set the requirements, prioritize it through the product team, evaluate, schedule, implement. Unless we're talking simply about copying and pasting two lines of code from RBNO to Raft branch, which, as I wrote above, would only "fix"  a very narrow window,  it'a sizable effort.

**Comment by kostja** [comment]
I checked the original #11302. I don't see how the heck it was possible to accept this issue, review it, queue a patch (without a regression test - which would show immediately it's broken with Raft, in August 2022 when we were already full throttle working on consistent topology management which is focused squarely on this class of problems. 

I think you can safely reopen this issue - there is a ton of cases when it's still unfixed. Just running decommission immediately after iptables and before gossip got a chance to notice it would reproduce it.

**Comment by fruch** [comment]
@kostja I'll leave it for the guys on the master call next week to decide on how to roll on with this topic (i.e. as a new issue or reopening the old one)

**Comment by mykaul** [comment]
@kostja - what's the next step here?

**Comment by kostja** [comment]
We may consider implementing it on the strongly consistent topology changes track but only after all the main cases are taken care of. 
Right now there is no design, and it will be a fairly complex feature. (1-3 man/month for a senior dev). 
If you believe it's part of the release criteria for the strongly consistent topology changes, I'll add it to the list.

As I wrote earlier it's not a regression - the feature is not implemented - it is made to work in the narrow window covered by the test. Our QA should crack it open and show it's not working properly.

**Comment by kostja** [comment]
It now sometimes works.

**Comment by kbr-scylla** [comment]
@mykaul the issue is in my "post-release" backlog, marked for 6.1, it's not a blocker for 6.0. It is not a priority.

**Comment by mykaul** [comment]
> @mykaul the issue is in my "post-release" backlog, marked for 6.1, it's not a blocker for 6.0. It is not a priority.

I'm more concerned that it is marked as a feature.

**Comment by kostja** [comment]
@mykaul why?

**Comment by mykaul** [comment]
> @mykaul why?
Scope is less clear to me.
